engine/src/main/scala/cromwell/engine/workflow/workflowstore/WorkflowStoreHeartbeatWriteActor.scala

```diff
@@ -38,18 +52,60 @@ case class WorkflowStoreHeartbeatWriteActor(workflowStoreAccess: WorkflowStoreAc
   override def commandToData(snd: ActorRef): PartialFunction[Any, (WorkflowId, OffsetDateTime)] = {
     case command: WorkflowStoreWriteHeartbeatCommand => (command.workflowId, command.submissionTime)
   }
+
+  /*
+  WARNING: Even though this is in an actor, the logic deals with instances of Future that could complete in _any_ order,
+  and even call this method at the same time from different threads.
+
+  We are expecting the underlying FSM to ensure that the call to this method does NOT occur in parallel, waiting for
+  the call to `process` to complete.
+   */
+  private def trackRepeatedFailures(startTime: OffsetDateTime, workflowCount: Int)(processTry: Try[Int]): Unit = {
+    processTry match {
+      case Success(_) =>
+        firstFailureOption = None
+      case Failure(_: Exception) =>
+        if (firstFailureOption.isEmpty) {
+          firstFailureOption = Option(startTime)
+        }
+
+        for {
+          failureShutdownDuration <- failureShutdownDurationOption
+          firstFailure <- firstFailureOption
+        } {
+          val now = OffsetDateTime.now()
+          val failureDuration = JDuration.between(firstFailure, now)
+          if (failureDuration.toNanos > failureShutdownDuration.toNanos) {
```

Comment:
As I understand it the logic for shutting down:

1. we are waiting for multiple failures
1. a duration must have passed between the first and the last failure, and all results in between must have been failures.

If that is true, I see a couple issues:

* We should fail on the first error
* Even if we decide to wait and see if errors are transient (worth discussing), the current logic won't execute if the `failureShutdownDurationOption` isn't set.  Thus this setting shouldn't be optional.